cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemotePluginConfigurer.java

```diff
@@ -0,0 +1,88 @@
+ *
+ */
+
+package io.cdap.cdap.internal.app;
+
+import com.google.common.collect.Maps;
+import io.cdap.cdap.api.artifact.ArtifactScope;
+import io.cdap.cdap.api.plugin.Plugin;
+import io.cdap.cdap.api.plugin.PluginClass;
+import io.cdap.cdap.api.plugin.PluginProperties;
+import io.cdap.cdap.api.plugin.PluginSelector;
+import io.cdap.cdap.common.io.Locations;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactDescriptor;
+import io.cdap.cdap.internal.app.runtime.artifact.Artifacts;
+import io.cdap.cdap.internal.app.runtime.artifact.PluginFinder;
+import io.cdap.cdap.internal.app.runtime.plugin.FindPluginHelper;
+import io.cdap.cdap.internal.app.runtime.plugin.PluginInstantiator;
+import io.cdap.cdap.internal.app.runtime.plugin.PluginNotExistsException;
+import io.cdap.cdap.proto.id.ArtifactId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import org.apache.twill.filesystem.Location;
+
+import java.io.File;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * Remote plugin configurer when the program is run in ISOLATED mode
+ */
+public class RemotePluginConfigurer extends DefaultPluginConfigurer {
+  private final Map<String, Plugin> localPlugins;
+  private final File localPluginDir;
+
+  public RemotePluginConfigurer(ArtifactId artifactId, NamespaceId pluginNamespaceId,
+                                PluginInstantiator pluginInstantiator, PluginFinder pluginFinder,
+                                Map<String, Plugin> localPlugins, String localPluginDir) {
+    super(artifactId, pluginNamespaceId, pluginInstantiator, pluginFinder);
+    this.localPlugins = localPlugins;
+    this.localPluginDir = new File(localPluginDir);
+  }
+
+  @Override
+  protected Plugin addPlugin(String pluginType, String pluginName, String pluginId,
+                             PluginProperties properties, PluginSelector selector) throws PluginNotExistsException {
+    validateExistingPlugin(pluginId);
+    // this means this plugin is not registered when the app is initially deployed, use the normal way to find the
+    // plugin
+    if (!localPlugins.containsKey(pluginId)) {
+      return super.addPlugin(pluginType, pluginName, pluginId, properties, selector);
+    }
+
+    Plugin existingPlugin = localPlugins.get(pluginId);
+    File existingPluginLocation = new File(localPluginDir, Artifacts.getFileName(existingPlugin.getArtifactId()));
+
+    // need to regenerate this plugin to ensure the plugin has updated properties with macro resolved, also
+    // register it to plugin instantiator
+    io.cdap.cdap.api.artifact.ArtifactId artifactId = existingPlugin.getArtifactId();
+    String namespace = artifactId.getScope().equals(ArtifactScope.SYSTEM) ? NamespaceId.SYSTEM.getNamespace() :
+                         pluginNamespaceId.getNamespace();
+    Location pluginLocation = Locations.toLocation(existingPluginLocation);
+
+    SortedMap<io.cdap.cdap.api.artifact.ArtifactId, PluginClass> selectedPlugin = new TreeMap<>();
```

Comment:
Just out of general curiosity, what is the reason for using a sorted map in here?

Reply:
`ArtifactId` is a comparable, this is to ensure when the first entry of the map is the latest, I believe our plugin selector always choose the first one if version is not specified.